### ISSUE TITLE
Evolve template data for usability!

### ISSUE THREAD

**Issue by christophermaier** [habitat-sh/habitat#4872]
Now that our templating data are cleanly separated from the internal data structures that feed them (see  #4823), we have the opportunity to start changing what we expose in order to make templates more expressive and easier to use.

A few ideas to start with:
* Expose the leader of a service group explicitly (e.g. `{{bind.<SERVICE>.leader}}`. If not in a leader topology, this will be `null` (already done in #4836 😉)
* Expose the followers in a service group explicitly (e.g. `{{bind.<SERVICE>.followers}}`. If not in a leader topology, this will be an empty array.
* Instead of exposing all members of a service group (which includes confirmed-dead and departed members), expose only the "active" members (those that are alive or suspect), e.g. `{{bind.<SERVICE>.active_members}}`. The old `.members` array would be retained for backwards-compatibility and removed at a future date. Furthermore, every other array of service members mentioned here would also contain just active members.
* Explicitly order arrays of service members. Right now, I believe they are ordered from oldest to youngest, though more by accident of implementation than by intention. This chronological ordering may be fine (other aspects may also be useful), but formally committing to an ordering will be helpful for users, making it easier to reason about templating.  Providing easy direct access to the extremes of these lists will also be useful, so you can easily get at the oldest or most recent member, for example.
* A helper function to explicitly provide access to a random member of a group may also be useful.

In addition to explicit access to things like `leader` and `follower`, it can also be useful to provide some generic "do what I mean" accessors when you don't necessarily care about the topology of a service. For example (these names are just a straw man!), a `{{bind.<SERVICE>.distinguished_member}}` field could provide access to the leader, if running in a leader topology, or access to the oldest (or youngest, or ...) member of the group if running in a standalone topology. This would allow users to write concise templates when they don't particularly care about the topology of a service, but would still provide explicit access to topology-specific information in cases where they _do_ care.

This is not an exhaustive list; I'm curious for feedback from other Habitat users as well. What do you wish was accessible in Habitat templates that isn't there already, or is difficult or impossible to access?

Thanks to @irvingpop, who provided lots of good feedback in the development of this issue.

**Comment by reset** [comment]
@christophermaier I'd like to see `members` eventually contain everything except for confirmed dead members. I think this is what people actually are asking for when they say "give me all members". This is because "alive" doesn't mean it's a healthy service. It just means we know that the Supervisor is still around. If I know nobody is coming back, I really don't want to see those members in my config. So departed and confirmed dead are almost always what I don't want to see.

I wrote the helpers to allow you to pull what you needed out of the members array as a compromise to not breaking older packages which contain a service. If I hadn't had that worry I probably would have just done the above ^

**Comment by christophermaier** [comment]
@reset Point of clarification: wouldn't we want `members` to include both alive and suspect members, since the whole reason for the suspicion mechanism is to prevent changes due to false positives because of transient network hiccups?

In general, I agree that when you ask for "members", you likely don't care about confirmed dead and departed members (I cannot think of a time where you _would_ care about those, really). The suggestion above of exposing an `active_members` list was really just a purely-additive way to address this, but changing `members` directly would probably work here, given that people using it are probably already filtering out the dead members anyway.

**Comment by reset** [comment]
@christophermaier yes, I should have been more clear instead of saying "everything except confirmed dead" 😄.

Do not include *confirmed dead* and *departed*. All other states should appear. I'd black list instead of white list in case we include additional states. We should be aware of this being a potential side effect when/if adding any additional health states.

**Comment by christophermaier** [comment]
Addendum: it also may be worth having `{{bind.<SERVICE>}}` be non-null only if there are alive-or-suspect members. You could achieve the same thing by using, say `{{ #if bind.<SERVICE>.members }}`, but it may be more expressive to be able to use the shorter form. Just an idea.

**Comment by baumanj** [comment]
Valid as of v0.57
